mongocrypt/src/error.rs

```diff
@@ -0,0 +1,189 @@
+    ptr,
+};
+
+use mongocrypt_sys as sys;
+
+#[derive(Debug)]
+pub struct Error {
+    pub kind: ErrorKind,
+    pub code: u32,
+    pub message: Option<String>,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?} ({})", self.kind, self.code)?;
+        if let Some(s) = &self.message {
+            write!(f, ": {}", s)?;
+        }
+        Ok(())
+    }
+}
+
+impl std::error::Error for Error {}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum ErrorKind {
+    // These correspond to errors from libmongocrypt
+    Client,
+    Kms,
+    CsFle,
+    // These are produced in this crate
+    Encoding,
+    Overflow,
+    Internal,
+}
+
+impl From<std::num::TryFromIntError> for Error {
+    fn from(err: std::num::TryFromIntError) -> Self {
+        overflow!("size overflow: {}", err)
+    }
+}
+
+impl From<std::str::Utf8Error> for Error {
+    fn from(err: std::str::Utf8Error) -> Self {
+        encoding!("invalid string: {}", err)
+    }
+}
+
+macro_rules! internal {
+    ($($arg:tt)*) => {{
+        crate::error::Error {
+            kind: crate::error::ErrorKind::Internal,
+            code: 0,
+            message: Some(format!($($arg)*)),
+        }
+    }}
+}
+pub(crate) use internal;
+
+macro_rules! encoding {
+    ($($arg:tt)*) => {{
+        crate::error::Error {
+            kind: crate::error::ErrorKind::Encoding,
+            code: 0,
+            message: Some(format!($($arg)*)),
+        }
+    }}
+}
+pub(crate) use encoding;
+
+macro_rules! overflow {
+    ($($arg:tt)*) => {{
+        crate::error::Error {
+            kind: crate::error::ErrorKind::Overflow,
+            code: 0,
+            message: Some(format!($($arg)*)),
+        }
+    }}
+}
+pub(crate) use overflow;
+
+use crate::{convert::str_bytes_len, native::OwnedPtr};
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+pub(crate) struct Status {
+    inner: OwnedPtr<sys::mongocrypt_status_t>,
+}
+
+impl Status {
+    pub(crate) fn new() -> Self {
+        Self::from_native(unsafe { sys::mongocrypt_status_new() })
+    }
+
+    pub(crate) fn from_native(inner: *mut sys::mongocrypt_status_t) -> Self {
+        Self {
+            inner: OwnedPtr::new(inner, sys::mongocrypt_status_destroy),
+        }
+    }
+
+    pub(crate) fn native(&self) -> &*mut sys::mongocrypt_status_t {
+        self.inner.borrow()
+    }
+
+    pub(crate) fn set(&mut self, err: &Error) -> Result<()> {
+        let inner_err;
+        // Reborrow to help the borrow checker accept the lifetime of the assignment in the fallthrough.
+        let mut err = &*err;
+        let typ = match err.kind {
+            ErrorKind::Client => sys::mongocrypt_status_type_t_MONGOCRYPT_STATUS_ERROR_CLIENT,
+            ErrorKind::Kms => sys::mongocrypt_status_type_t_MONGOCRYPT_STATUS_ERROR_KMS,
+            ErrorKind::CsFle => sys::mongocrypt_status_type_t_MONGOCRYPT_STATUS_ERROR_CSFLE,
+            _ => {
```

Comment:
Error handling in this crate is a complicated situation.  For errors returned directly by the wrapper, the user will be able to see all of the crate-specific error types.

However, because errors returned by callbacks have to transit via a native `Status` object, they can't accommodate those extra types, so if a callback returns one of those it's massaged via the code above into `ErrorKind::Client`.  I did experiment with having a distinct error that's returned by callbacks that doesn't have the wrapper error types, but that felt like it was adding more noise than utility.  Can switch back to that if people have strong opinions :)